Deprecate succeedLazy

ioleo opened the pull request:
Fixes #1346

* Deprecated all `succeedLazy` methods.
* Delegated all deprecated methods to `succeed`. 
* Added missing `succeed` to `ZSink`, `Sink` and `ZStreamChunk`.
* Removed `succeedLazy` test from `RTSSpec`
* Replaced `succeedLazy` with `effect` or `effectTotal` where laziness was essential and with  `succeed` where lazyness was not required
* Updated docs

shankarshastri commented on docs/overview/creating_effects.md:
````diff
@@ -23,16 +23,16 @@ You can also create use methods in the companion objects of the `ZIO` type alias
 ```
 
-The `succeed` method is eager, which means the value passed to `succeed` will be constructed _before_ the method is invoked. Although this is the most common way to construct a successful effect, you can achieve lazy construction using the `ZIO.succeedLazy` method:
+The `succeed` method is eager, which means the value passed to `succeed` will be constructed _before_ the method is invoked. Although this is the most common way to construct a successful effect, you can achieve lazy construction using the `ZIO.succeed` method:
````
Lazy construction?

shankarshastri commented on docs/datatypes/io.md:
````diff
@@ -13,22 +13,16 @@ A value of type `IO[E, A]` describes an effect that may fail with an `E`, run fo
+val value: UIO[String] = IO.succeed("Hello World")
 ```
 
 The constructor uses non-strict evaluation, so the parameter will not be evaluated until when and if the `IO` action is executed at runtime, which is useful if the construction is costly and the value may never be needed.
````
Non-Strict Evaluation?

shankarshastri commented on docs/overview/creating_effects.md:
````diff
@@ -23,16 +23,16 @@ You can also create use methods in the companion objects of the `ZIO` type alias
 ```
 
-The value inside a successful effect constructed with `ZIO.succeedLazy` will only be constructed if absolutely required.
+The value inside a successful effect constructed with `ZIO.succeed` will only be constructed if absolutely required.
````
Same here.

ioleo commented:
@shankarshastri Fixed docs.

kaishh commented:
> Delegated all deprecated methods to `succeed`.

I think we should delegate to `effectTotal` to not break existing usages and since `effectTotal` is now the only blessed way to delay evaluation.

kaishh commented on benchmarks/src/main/scala/zio/IODeepLeftBindBenchmark.scala:
```diff
@@ -34,9 +34,9 @@ class IODeepLeftBindBenchmark {
   def zioDeepLeftBindBenchmark(runtime: Runtime[Any]): Int = {
     var i  = 0
-    var io = IO.succeedLazy(i)
+    var io = IO.succeed(i)
```
This should become `effectTotal` in line with other effect types that delay execution in their benchmarks (e.g. monix `Task.eval`)

ioleo replied:
Fixed

kaishh commented on benchmarks/src/main/scala/zio/IODeepFlatMapBenchmark.scala:
```diff
@@ -122,10 +122,10 @@ class IODeepFlatMapBenchmark {
   private[this] def zioDeepFlatMap(runtime: Runtime[Any]): BigInt = {
     def fib(n: Int): UIO[BigInt] =
-      if (n <= 1) ZIO.succeedLazy[BigInt](n)
+      if (n <= 1) ZIO.succeed[BigInt](n)
```
This should become `effectTotal` in line with other effect types that delay execution in their benchmarks (e.g. monix `Task.eval`)

ioleo replied:
Fixed

kaishh commented on benchmarks/src/main/scala/zio/IOLeftBindBenchmark.scala:
```diff
@@ -120,11 +120,11 @@ class IOLeftBindBenchmark {
-      if (i % depth == 0) IO.succeedLazy[Int](i + 1).flatMap(loop)
-      else if (i < size) loop(i + 1).flatMap(i => IO.succeedLazy(i))
-      else IO.succeedLazy(i)
+      if (i % depth == 0) IO.succeed[Int](i + 1).flatMap(loop)
```
This should become `effectTotal` in line with other effect types that delay execution in their benchmarks (e.g. monix `Task.eval`)

ioleo replied:
:heavy_check_mark:

kaishh commented on benchmarks/src/main/scala/zio/IOMapBenchmark.scala:
```diff
@@ -113,7 +113,7 @@ class IOMapBenchmark {
       else sumTo(t.map(_ + n), n - 1)
 
-    runtime.unsafeRun(sumTo(IO.succeedLazy(0), depth))
+    runtime.unsafeRun(sumTo(IO.succeed(0), depth))
```
This should become `effectTotal` in line with other effect types that delay execution in their benchmarks (e.g. monix `Task.eval`)

ioleo replied:
:heavy_check_mark:

kaishh commented on benchmarks/src/main/scala/zio/IONarrowFlatMapBenchmark.scala:
```diff
@@ -112,10 +112,10 @@ class IONarrowFlatMapBenchmark {
     def loop(i: Int): UIO[Int] =
-      if (i < size) IO.succeedLazy[Int](i + 1).flatMap(loop)
-      else IO.succeedLazy(i)
+      if (i < size) IO.succeed[Int](i + 1).flatMap(loop)
```
This should become `effectTotal` in line with other effect types that delay execution in their benchmarks (e.g. monix `Task.eval`)

ioleo replied:
:heavy_check_mark:

kaishh commented on benchmarks/src/main/scala/zio/IOShallowAttemptBenchmark.scala:
```diff
@@ -130,7 +130,7 @@ class IOShallowAttemptBenchmark {
     def throwup(n: Int): IO[ScalazError, BigInt] =
       if (n == 0) throwup(n + 1).fold[BigInt](_ => 50, identity)
-      else if (n == depth) IO.succeedLazy(1)
+      else if (n == depth) IO.succeed(1)
```
This should become `effectTotal` in line with other effect types that delay execution in their benchmarks (e.g. monix `Task.eval`)

ioleo replied:
:heavy_check_mark:

kaishh commented on core-tests/jvm/src/test/scala/zio/IOSpec.scala:
```diff
@@ -122,19 +122,19 @@ class IOSpec(implicit ee: org.specs2.concurrent.ExecutionEnv) extends TestRuntim
   def t3 = {
     val list = List("1", "h", "3")
-    val res  = Try(unsafeRun(IO.foreach(list)(x => IO.succeedLazy[Int](x.toInt))))
+    val res  = Try(unsafeRun(IO.foreach(list)(x => IO.effect[Int](x.toInt))))
```
I think all `_.toInt` in this file should become either `effectTotal` or `succeed` consistently (since it's all the same call and there's no difference between them)

ioleo replied:
Actually I did initially try with `succeed`, but the test failed (aparently laziness matters here).

ioleo replied:
In `t3` lazyness matters, so changing all `_.toInt` to `effectTotal`.

kaishh commented on core-tests/jvm/src/test/scala/zio/IOSpec.scala:
```diff
@@ -144,26 +144,26 @@ class IOSpec(implicit ee: org.specs2.concurrent.ExecutionEnv) extends TestRuntim
 
   def t7 = {
-    val list = List(1, 2, 3).map(IO.succeedLazy[Int](_))
+    val list = List(1, 2, 3).map(IO.succeed[Int](_))
```
I think it's better to replace most of these by `effectTotal` since otherwise the fiber workload becomes too trivial which may interfere with the original intention of the test writer...

ioleo replied:
:heavy_check_mark:

kaishh commented on core/shared/src/main/scala/zio/Fiber.scala:
```diff
@@ -285,11 +285,9 @@ object Fiber {
-   * evaluated value.
-   */
-  final def succeedLazy[E, A](a: => A): Fiber[E, A] = done(Exit.succeed(a))
+  @deprecated("use succeed", "1.0.0")
```
There is no `effectTotal` on `Fiber`, so I'm inclined to let this one stay since there's no alternative?.. /cc @jdegoes

vasilmkd commented:
What is the impact of this on `ZStream`, given that `ZManaged.succeedLazy` is used to implement effectful folding? /cc @iravid

iravid commented:
`ZManaged.succeedLazy(a)` is just `ZManaged.fromEffect(ZIO.effectTotal(a))` so we can name that `ZManaged.effectTotal`. Can be done in a separate PR as other data types have a `succeedLazy` too.

ioleo commented:
@kaishh 
There is no `effectTotal` on `Fiber`, `STM`, `Schedule` and `ZSchedule`. For all others I've switched to `effectTotal` as you sugested. For `Managed` / `ZManaged` I introduced `effectTotal` as @iravid suggested.

jdegoes commented:
@ioleo Excellent work, thank you for this! 🙏
